pallets/pooled-staking/src/migrations.rs

```diff
@@ -0,0 +1,80 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Tanssi.  If not, see <http://www.gnu.org/licenses/>
+
+use crate::{
+    Candidate, CandidateSummaries, CandidateSummary, Config, Delegator,
+    DelegatorCandidateSummaries, DelegatorCandidateSummary, PoolKind, Pools, PoolsKey,
+};
+use frame_support::pallet_prelude::Zero;
+use sp_core::Get;
+use sp_runtime::Weight;
+use sp_std::collections::btree_map::BTreeMap;
+
+pub fn generate_summaries<T: Config>(_available_weight: Weight) -> Weight {
+    let mut delegator_summaries =
+        BTreeMap::<(Delegator<T>, Candidate<T>), DelegatorCandidateSummary>::new();
+    let mut candidate_summaries = BTreeMap::<Candidate<T>, CandidateSummary>::new();
+
+    let mut read_count = 0;
+    let mut write_count = 0;
+
+    for (candidate, key, value) in Pools::<T>::iter() {
+        read_count += 1;
+
+        let (pool, delegator) = match key {
+            PoolsKey::JoiningShares { delegator } => (PoolKind::Joining, delegator),
+            PoolsKey::AutoCompoundingShares { delegator } => (PoolKind::AutoCompounding, delegator),
+            PoolsKey::ManualRewardsShares { delegator } => (PoolKind::ManualRewards, delegator),
+            PoolsKey::LeavingShares { delegator } => (PoolKind::Leaving, delegator),
+            _ => continue, // we only care about share values
+        };
+
+        // Are 0/Default values automatically from storage?
+        // In case they aren't we check the amount of shares isn't 0.
+        if value.is_zero() {
+            continue;
+        }
+
+        let summary = delegator_summaries
+            .entry((delegator, candidate.clone()))
+            .or_default();
+        summary.set_pool(pool, true);
+
+        let summary = candidate_summaries.entry(candidate.clone()).or_default();
+        let pool_count = summary.pool_delegators_mut(pool);
+        *pool_count = pool_count.saturating_add(1);
+    }
+
+    // We now iterate over delegator_summaries to fill candidate_summaries with delegators
+    // count without risking having duplicates due to multiple pools.
+    for (_delegator, candidate) in delegator_summaries.keys() {
+        let summary = candidate_summaries.entry(candidate.clone()).or_default();
+        summary.delegators = summary.delegators.saturating_add(1);
+    }
+
+    // Now that we have collected all info, we put them in real storage.
+    for ((delegator, candidate), summary) in delegator_summaries.into_iter() {
+        write_count += 1;
+        DelegatorCandidateSummaries::<T>::insert(delegator.clone(), candidate.clone(), summary);
+    }
+
+    for (candidate, summary) in candidate_summaries.into_iter() {
+        write_count += 1;
+        CandidateSummaries::<T>::insert(candidate, summary);
+    }
+
+    T::DbWeight::get().reads_writes(read_count, write_count)
```

Comment:
What would be the easiest way to do this migration if it uses more than one block of weight? Any ideas?

Reply:
Iterate over `Pools` entries, accumulate the numbers and when we run out of available weight we store as migration state the output of `iterator.last_raw_key()`, which can be used to resume the iteration in the next block. 

I just need to rewrite the migration code in a way that doesn't need to accumulate in a big BTreeMap, and instead probably directly read/write from storage. For a candidate delegators count (any pool without duplicate), I should be able to do it without the BTreeMap by increasing this number only if the delegator summary goes from empty to some value.

Reply:
Ok sounds hard. Nothing will break if the pools are updated while the migration is running?

Reply:
Extrinsics needs to be paused.
Rewarding can change amounts, but not change value from/to zero; which is what counts for the summaries, so we don't need to pause rewarding.